ssl/quic/quic_thread_assist.c

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright 2023 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/macros.h>
+#include "quic_local.h"
+#include "internal/time.h"
+#include "internal/thread.h"
+#include "internal/thread_arch.h"
+#include "internal/quic_thread_assist.h"
+
+#if !defined(OPENSSL_NO_QUIC) && !defined(OPENSSL_NO_THREAD_POOL)
+
+/* Main loop for the QUIC assist thread. */
+static unsigned int assist_thread_main(void *arg)
+{
+    QUIC_THREAD_ASSIST *qta = arg;
+    CRYPTO_MUTEX *m = ossl_quic_channel_get_mutex(qta->ch);
+    QUIC_REACTOR *rtor;
+
+    if (!CRYPTO_THREAD_write_lock(m))
+        return 0;
+
+    rtor = ossl_quic_channel_get_reactor(qta->ch);
+
+    for (;;) {
+        if (qta->teardown)
+            break;
+
+        ossl_crypto_condvar_wait_timeout(qta->cv, m,
+                                         ossl_quic_reactor_get_tick_deadline(rtor));
+
+        /*
+         * We have now been woken up. This can be for one of the following
+         * reasons:
+         *
+         *   - We have been asked to teardown (qta->teardown is set);
+         *   - The tick deadline has passed.
+         *   - The tick deadline has changed.
+         *
+         * For robustness, this loop also handles spurious wakeups correctly
+         * (which does not require any extra code).
+         */
+        if (qta->teardown)
+            break;
+
+        ossl_quic_reactor_tick(rtor);
+    }
+
+    CRYPTO_THREAD_unlock(m);
+    return 1;
+}
+
+int ossl_quic_thread_assist_init_start(QUIC_THREAD_ASSIST *qta,
+                                       QUIC_CHANNEL *ch)
+{
+    CRYPTO_MUTEX *mutex = ossl_quic_channel_get_mutex(ch);
+
+    if (mutex == NULL)
+        return 0;
+
+    qta->ch         = ch;
+    qta->teardown   = 0;
+    qta->joined     = 0;
+
+    qta->cv = ossl_crypto_condvar_new();
+    if (qta->cv == NULL)
+        return 0;
+
+    qta->t = ossl_crypto_thread_native_start(assist_thread_main,
```

Comment:
This mixes abstraction layers -- thread pool and arch-specific implementation:
 - there's a thread pool preprocessor check on line 17 (`!defined(OPENSSL_NO_THREAD_POOL)`), but
 - this call is thread pool agnostic, will work even if there's no support for thread pool and circumvents thread pool limit checks.

Not sure if this was intended; perhaps using `defined(OPENSSL_THREADS)` on line 17 is a better fit?